packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -711,8 +713,11 @@ Hardhat Network's forking functionality only works with blocks from at least spu
       txPoolSnapshotId: this._txPool.snapshot(),
       blockTimeOffsetSeconds: this.getTimeIncrement(),
       nextBlockTimestamp: this.getNextBlockTimestamp(),
+      irregularStatesByBlockNum: this._irregularStatesByBlockNum,
     };
 
+    this._irregularStatesByBlockNum = new Map(this._irregularStatesByBlockNum);
```

Comment:
It clones the map so that the post-snapshot environment will have its own copy.  This way, any irregular states that come in while the snapshot is in effect will be applied only to the map that's specific to that snapshot.  When the snapshot is reverted, the map referred to by the snapshot object will again be used, and the clone will be garbage collected since the variable that was referring to it will then be referring to the original map that was in use before the snapshot was taken.

I found that, without this line, when a snapshot was in effect, any changes to the map would remain in the map even after the snapshot was reverted, because it was all operating on the same map.  At first I tried to use a simple assignment rather than `new Map()`, but it still didn't work, and I concluded that that was because a simple assignment was only creating an additional reference to the same object instance, and changing to `new Map()` made everything work nicely, so I assumed that this validated that conclusion.

Reply:
TIL: you can clone a map with `new Map(map)`. I used to do `new Map(map.entries())`

Agreed that cloning is necessary. Some time ago this method cloned lots of things like that.